pip/pip-385.md

```diff
@@ -0,0 +1,348 @@
+established connection for a topic. This is transparent to the clients, and they continue to publish further messages
+with an increased observed latency. Once the publish-rates are within the limits, broker resumes reading from the
+connection.
+
+Here is a small illustration to demonstrate the situation:
+
+```mermaid
+%%{init: {"mirrorActors": false, "rightAngles": false} }%%
+sequenceDiagram
+    Client->>Server: CreateProducer(reqId, myTopic)
+    Note right of Server: Check Authorization
+    Server-->>Client: ProducerSuccess(reqId, producerName)
+    Activate Client
+    Activate Server
+    Client->>Server: Send(1, message1)
+    Client->>Server: Send(2, message2)
+    Server-->>Client: SendReceipt(1, msgId1)
+    Client->>Server: Send(3, message3)
+    Client->>Server: Send(4, message4)
+    Note right of Server: Topic breaching quota
+    Activate Server
+    note right of Server: TCP channel read paused
+    Client-xServer: Send(5, message5)
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Client-xServer: Send(6, message6)
+    Server-->>Client: SendReceipt(4,msgId4)
+    note right of Server: TCP channel read resumed
+    deactivate Server
+    Server-->>Server: read message 5
+    Server-->>Server: read message 6
+    Client->>Server: Send(7, message7)
+    Server-->>Client: SendReceipt(5,msgId5)
+    Server-->>Client: SendReceipt(6,msgId6)
+    Server-->>Client: SendReceipt(7,msgId7)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Challenges with the current approach
+
+The current approach may look perfectly fine when looking at the above example, but when looked from a wider scope,
+things start looking bad.
+Typically, the clients reuse a single TCP connection from the client to a broker to send messages to multiple topics.
+This is controlled by the client side property
+of [connectionsPerBroker](https://github.com/apache/pulsar/blob/4b3b273c1c57741f9f9da2118eb4ec5dfeee2220/pulsar-client/src/main/java/org/apache/pulsar/client/impl/conf/ClientConfigurationData.java#L135)
+which defaults to 1. The situation is worsened by the fact that typically, a client is used to create producers for
+partitioned topics and generally an application may produce to more than one partitioned topic with the producers
+created from the same client object, thus all sharing the same tcp connection.
+
+In this situation, even when a single topic starts breaching the quota, the entire TCP connection is paused leading to
+a noisy neighbour effect where effectively all the topics that the client is producing to start getting throttled and
+observe high latencies.
+
+# Motivation
+
+The current method of inducing throttling when a topic or connection breaches quota has various challenges:
+
+* **Noisy neighbors** - Even if one topic is exceeding the quota, since the entire channel read is paused, all topics
```

Comment:
I don't think it's a noisy neighbor issue. There are many ways: clients can use a separate connection for different topics by increasing the number of connections and more specifically create Cache of PulsarClient objects to manage topics belonging to different usecases. If you use one channel for different tenants/usecases and if they get impacted then it's not a noisy neighbor but the application might need design improvement. 
For example: If client app use the same topic for different usecases then all usecases can be impacted by each other, and that doesn't mean Pulsar has a noisy neighbor issue but it needs a design change to use separate topics for each usecase. So, this challenge is easily achievable.

Reply:
Hi Rajan, thanks for going through the proposal. I think your workarounds and the discussions that will follow would be a good addition to the PIP for the alternate/challenges section. Replies below.

> For example: If client app use the same topic for different usecases then all usecases can be impacted by each other, and that doesn't mean Pulsar has a noisy neighbor issue but it needs a design change to use separate topics for each usecase

I agree that this particular example needs a application level redesign to separate out the use cases into different topics. That's not part of the problem statement here.

> clients can use a separate connection for different topics by increasing the number of connections and more specifically create Cache of PulsarClient objects to manage topics belonging to different usecases

From running a messaging (platform as a service) in a big organization for over 2 years using pulsar, we have tried a lot of things to overcome the challenges I've listed, including what you are suggesting here.
There are many challenges with this approach:

* On the client side, a topic may have 100s of partitions and a single application may be producing to 2-3 such partitioned topics where the topics cannot be further split. At this point, the client pod is limited in resources and spinning 1 TCP connection per partition might not be feasible from a network point of view.
* Having 1 cached client per topic (and essentially per partition here) leads to other issue - properties like numListenerThreads, numIoThreads, memoryLimitBytes etc loose their effectiveness.
* On the broker side, assuming a broker is owner of 2000 topics.. the broker now has to handle many many TCP connections. probably in the range of 20k-100k. There is a network and OS level limit to how many connections can be present and moreover - a broker's performance varies a lot based on how many actual connections are open to it. We have tested this in NFR/redline that performance degrades as the connections go up.

Moreover, from a design perspective, it is very clear that the pulsar protocol and the java client implementation is designed keeping shared/multiple topics in a single connection in mind. This looks like one of the core fundamentals of the binary protocol as well - thus, the producer id is present in almost all commands. Now if this is the core construct of the design, then it should guarantee zero noise as well. Users shouldn't have to hack their way around this.

> If you use one channel for different tenants/usecases and if they get impacted then it's not a noisy neighbor but the application might need design improvement.

There is no limit to this. Noisy neighbor can happen within a single tenant as well. Within the same application and tenant itself, some topics might be holding analytical data and some topics might be holding transactional data.. it doesn't seem right to block the transactional topics when the analytical ones are breaching the quota.


Having said all of the above, do you see any particular reason why this improvement should NOT land in pulsar?
Its not the first time that improvements are being proposed for things that were possible previously, but with either some repetitive action or some outside effort - PIP-352, 356 and 383 are some of the recent PIPs that come to my mind.

Reply:
@rdhabalia - gentle reminder . Would love to get a closure here.